### DISCUSSION TITLE
submit Block issue

### DISCUSSION THREAD

**Discussion by jack8086wang** [nervosnetwork/ckb#2415] [General]
1)校验算法：https://github.com/leifjacky/ckb-gominer-demo/tree/master/eaglesong/src
2) 用这个demo里面的headerhash+nonce1+nonce2 测试过校验算法没问题：https://github.com/leifjacky/ckb-gominer-demo

**Comment by jack8086wang** [comment]
3）用测试节点测试submit block时报header(pow(invalidnonce))
以下是get_block_template和submit block的数据：
----------------------------------------------------------------
```
template block:  {"bytes_limit":"0x91c08","cellbase":{"cycles":null,"data":{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x30f68"}],"outputs":[{"capacity":"0x4d8d068e68","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x5a0000000c0000005500000049000000100000003000000031000000d41a05bdd8e5bb6b236d9399d08c543a159e44762466494d0fb48ea74bfccd560114000000fce926a49302c3f30f4fb56fa683a3f2bc60fb650100000000"]},"hash":"0xd5a4b26101948d592b10b1427b7fd7ef785c1d4037f4f28a2e164f42c9b9ceb9"},"compact_target":"0x1cc5f5cb","current_time":"0x171ff1b5572","cycles_limit":"0xd09dc300","dao":"0x59dd4f741fb9932f2423d0876ab32300eb87bd67300d320000aebe36802d0707","epoch":"0x25b01c300010d","number":"0x30f68","parent_hash":"0x313d112e4db087fc06c309458feb178b61bbea3e41ca08a73c0b699db6f2b884","proposals":[],"transactions":[],"uncles":[],"uncles_count_limit":"0x2","version":"0x0","work_id":"0x15e","height":200552,"difficulty":482735563}
work_id =  0x15e
rpc params =  ["0x15e",{"header":{"version":"0x0","compact_target":"0x1cc5f5cb","timestamp":"0x171ff1b5572","number":"0x30f68","epoch":"0x25b01c300010d","transactions_root":"0xf68ce59d643dc972042b9fa0aef1c90a8e2fdbb5a8c3c433eab4e0cbb5798faa","proposals_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","uncles_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","parent_hash":"0x313d112e4db087fc06c309458feb178b61bbea3e41ca08a73c0b699db6f2b884","dao":"0x59dd4f741fb9932f2423d0876ab32300eb87bd67300d320000aebe36802d0707","nonce":"0x56e261ca00000000000010b6c66eb"},"uncles":[],"transactions":[{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x30f68"}],"outputs":[{"capacity":"0x4d8d068e68","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x5a0000000c0000005500000049000000100000003000000031000000d41a05bdd8e5bb6b236d9399d08c543a159e44762466494d0fb48ea74bfccd560114000000fce926a49302c3f30f4fb56fa683a3f2bc60fb650100000000"]}],"proposals":[]}]
^_^_^_^_^_^_^_^: submit_block rpc result: error= { code: -3, message: 'Header(Pow(InvalidNonce))' } , result =  undefined
======================================================================
method =  mining.submit params =  ["ckb1qyqwy2tfk2h4s3w78xulyhh8xa2ymgj433qse68dqt.192X168X2x2","244917948241345","09500000000000015622c418"] ip =  110.87.91.88
target1:  35f5f1b20c96a626ddcbe2ef6b6e919fabc06c7950b7b6d78176af
target2:  01ffffffffffffffffffffffffffffffffffffffffffffffffffffff
target3:  c5f5cb00000000000000000000000000000000000000000000000000
record & submit to chain ...
2020-05-10 23:02:04 [pool] (Thread 2) Accepted  share at difficulty 2000/NaN from ckb1qyqwy2tfk2h4s3w78xulyhh8xa2ymgj433qse68dqt.192X168X2x2@110.87.91.88
---------------------------------------------------------------------
template block:  {"bytes_limit":"0x91c08","cellbase":{"cycles":null,"data":{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x30f68"}],"outputs":[{"capacity":"0x4d8d068e68","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x5a0000000c0000005500000049000000100000003000000031000000d41a05bdd8e5bb6b236d9399d08c543a159e44762466494d0fb48ea74bfccd560114000000fce926a49302c3f30f4fb56fa683a3f2bc60fb650100000000"]},"hash":"0xd5a4b26101948d592b10b1427b7fd7ef785c1d4037f4f28a2e164f42c9b9ceb9"},"compact_target":"0x1cc5f5cb","current_time":"0x171ff1b5572","cycles_limit":"0xd09dc300","dao":"0x59dd4f741fb9932f2423d0876ab32300eb87bd67300d320000aebe36802d0707","epoch":"0x25b01c300010d","number":"0x30f68","parent_hash":"0x313d112e4db087fc06c309458feb178b61bbea3e41ca08a73c0b699db6f2b884","proposals":[],"transactions":[],"uncles":[],"uncles_count_limit":"0x2","version":"0x0","work_id":"0x15e","height":200552,"difficulty":482735563}
work_id =  0x15e
rpc params =  ["0x15e",{"header":{"version":"0x0","compact_target":"0x1cc5f5cb","timestamp":"0x171ff1b5572","number":"0x30f68","epoch":"0x25b01c300010d","transactions_root":"0xf68ce59d643dc972042b9fa0aef1c90a8e2fdbb5a8c3c433eab4e0cbb5798faa","proposals_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","uncles_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","parent_hash":"0x313d112e4db087fc06c309458feb178b61bbea3e41ca08a73c0b699db6f2b884","dao":"0x59dd4f741fb9932f2423d0876ab32300eb87bd67300d320000aebe36802d0707","nonce":"0x56e2609500000000000015622c418"},"uncles":[],"transactions":[{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x30f68"}],"outputs":[{"capacity":"0x4d8d068e68","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x5a0000000c0000005500000049000000100000003000000031000000d41a05bdd8e5bb6b236d9399d08c543a159e44762466494d0fb48ea74bfccd560114000000fce926a49302c3f30f4fb56fa683a3f2bc60fb650100000000"]}],"proposals":[]}]
^_^_^_^_^_^_^_^: submit_block rpc result: error= { code: -3, message: 'Header(Pow(InvalidNonce))' } , result =  undefined
======================================================================
```
target1: powhash
target2: 下发矿机的target
target3: block_target
本地计算powhash<block.target, 但是submit 之后节点校验不通过，请帮忙确认会是什么可能

**Comment by doitian** [comment]
用的哪个版本？ v0.32.0-rc1 在使用 testnet spec 的清空下用了新的算法，和现在的算法是不兼容的。最新的 develop 分支上也是一样。

**Comment by doitian** [comment]
## Header

To ease PoW computation, the header is split into `raw` and `nonce`. 

```
struct Header {
    raw:                    RawHeader,
    nonce:                  Uint128,
}
```

The header must meet the last inequality in the following snippet:

```
pow_hash := ckbhash(molecule_serialize(raw))
pow_message := pow_hash || to_le(nounce)
pow_output := eaglesong(pow_message)

from_be(pow_output) <= compact_to_target(raw.compact_target)
```

Functions used in the pseudocode:

* `:=`: assignment
* `||`: binary concatenation.
* `ckbhash`: Blake2b hash with CKB specific configuration, see Appendix.
* `to_le`: Convert unsigned integer to bytes in little endian. The bytes count is the same with the integer width.
* `from_be`: Convert bytes encoded in big endian to an unsigned integer.
* `molecule_serialize`: Serialize a structure into binary using its schema.
* `eaglesong`: See #todo rfc
* `compact_to_target`: `raw.compact_target` encodes the difficulty target in a compact form. This function restores the target from the compact from.

The block is usually referenced by the header hash, for example, in `raw.parent_hash`.

```
header_hash := ckb_hash(molecule_serialize(header))
```

Notice that Header and RawHeader are all fixed size structure. The serialization of them are just the simple binary concatenation of the fields in order.

**Comment by doitian** [comment]
比较容易错的地方有两个

* nonce 是按小端序放入 eaglesong input 的，且 raw hash 在前，nonce 在后
* 和 target 比较的时候，eaglesong output 是按大端序转成整数的

**Comment by jack8086wang** [comment]
我用的版本是V0.31.1版本，所以这个版本上协议sumbit应该不是下列的方式吧
To ease PoW computation, the header is split into raw and nonce.

struct Header {
    raw:                    RawHeader,
    nonce:                  Uint128,
}

另外一个问题是，在最新版本v.32.0.rc-1上只有testnet采用plit into raw and nonce的方式， mainnet是还是沿用之前的结构。nonce包含在header里面，我的理解是只有在测试网的有变化？

**Comment by doitian** [comment]
0.32.0 testnet 改动是其它的，多加了一轮 blake2b。上面发的是相通的。

分成 raw 和 nonce 是序列化层面的，RPC 里是摊平的。

**Comment by jack8086wang** [comment]
那很奇怪，我试过header分成raw和nonce,    submit的时候会报下面的错, 所以我一直把nonce放在header里面，这个是哪里有问题：
rpc params =  ["0x15",{"header":{"raw":{"version":"0x0","compact_target":"0x1cbbeaf6","timestamp":"0x172065af9fd","number":"0x32492","epoch":"0x26d0080000116","transactions_root":"0x986168d60991ac43faf3bc76dd01f2276875076c857f09d2827cd6a251b3e77c","proposals_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","uncles_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","parent_hash":"0x1cf46db27018437b076a4a7f828a416fb20cfd21fda733727067ec12b1e7c279","dao":"0x074768ed84549b2f2c4db3b2ccb423001ec2e228919f33000055b632a44b0707"},"nonce":"0xd6fce02c000000000000277b01951"},"uncles":[],"transactions":[{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x32492"}],"outputs":[{"capacity":"0x4b4d16dfb3","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x5b0000000c00000055000000490000001000000030000000310000009f481f378dd1ef1384212aacaec3d3cb427213c68c9895cfb87efdfd8f7944f00114000000eacd8332032b70db4b07bf1a7eb62fe1994b2679020000008806"]}],"proposals":[]}]
^_^_^_^_^_^_^_^: submit_block rpc result: error= { code: -32602,
  message: 'Invalid params: unknown field `raw`, expected one of `version`, `compact_target`, `timestamp`, `number`, `epoch`, `parent_hash`, `transactions_root`, `proposals_hash`, `uncles_hash`, `dao`, `nonce`.' } , result =  undefined

**Comment by doitian** [comment]
> 0.32.0 testnet 改动是其它的，多加了一轮 blake2b。上面发的是相通的。 分成 raw 和 nonce 是序列化层面的，RPC 里是摊平的。

**RPC 里是摊平的**

**Comment by doitian** [comment]
我测试了你发的这个 request，PoW 验证是通过的。建议本地使用 v0.31.1 启动 dev chain 测试 `ckb init -c dev`

```
rpc params =  ["0x15e",{"header":{"version":"0x0","compact_target":"0x1cc5f5cb","timestamp":"0x171ff1b5572","number":"0x30f68","epoch":"0x25b01c300010d","transactions_root":"0xf68ce59d643dc972042b9fa0aef1c90a8e2fdbb5a8c3c433eab4e0cbb5798faa","proposals_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","uncles_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","parent_hash":"0x313d112e4db087fc06c309458feb178b61bbea3e41ca08a73c0b699db6f2b884","dao":"0x59dd4f741fb9932f2423d0876ab32300eb87bd67300d320000aebe36802d0707","nonce":"0x56e261ca00000000000010b6c66eb"},"uncles":[],"transactions":[{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x30f68"}],"outputs":[{"capacity":"0x4d8d068e68","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x5a0000000c0000005500000049000000100000003000000031000000d41a05bdd8e5bb6b236d9399d08c543a159e44762466494d0fb48ea74bfccd560114000000fce926a49302c3f30f4fb56fa683a3f2bc60fb650100000000"]}],"proposals":[]}]
```

**Comment by doitian** [comment]
> 我测试了你发的这个 request，PoW 验证是通过的。建议本地使用 v0.31.1 启动 dev chain 测试 `ckb init -c dev`

Sorry, dev chain 默认是用的 dummy，我等下把调试信息贴出来

使用 dev 测试的话，需要修改 specs/dev.toml，把最后一行改成

```
[pow]
func = "Eaglesong"
```

**Comment by doitian** [comment]
* pow hash: header 不包含 nonce 部分使用 molecule 序列化之后的数据
* nonce: 在 RPC 中是作为整数提交，但是要以小端序放入 pow input 中
* pow input: eaglesong 的输入，先放 pow hash，然后是小端序的 nonce
* pow output: eaglesong 的输出

```
pow hash: 0xada7e355e72bd62451f5e718a7bef920a8be641fda2a150f06a4c1645cfc734d
nonce: 0x56e261ca00000000000010b6c66eb
pow input: [
    0xad,
    0xa7,
    0xe3,
    0x55,
    0xe7,
    0x2b,
    0xd6,
    0x24,
    0x51,
    0xf5,
    0xe7,
    0x18,
    0xa7,
    0xbe,
    0xf9,
    0x20,
    0xa8,
    0xbe,
    0x64,
    0x1f,
    0xda,
    0x2a,
    0x15,
    0xf,
    0x6,
    0xa4,
    0xc1,
    0x64,
    0x5c,
    0xfc,
    0x73,
    0x4d,
    0xeb,
    0x66,
    0x6c,
    0xb,
    0x1,
    0x0,
    0x0,
    0x0,
    0x0,
    0x0,
    0xa0,
    0x1c,
    0x26,
    0x6e,
    0x5,
    0x0,
]
pow output: [
    0x98,
    0x17,
    0x8f,
    0x45,
    0x4c,
    0x3,
    0xef,
    0xde,
    0x7a,
    0x83,
    0x96,
    0xe4,
    0x46,
    0x99,
    0x6,
    0x8f,
    0x82,
    0xdb,
    0xae,
    0x3e,
    0xa1,
    0xe6,
    0x12,
    0x2e,
    0xf3,
    0x72,
    0x13,
    0xce,
    0x37,
    0x1d,
    0x21,
    0x8,
]
pow output as big endian: 0x98178f454c03efde7a8396e44699068f82dbae3ea1e6122ef37213ce371d2108
block_target: 0xc5f5cb00000000000000000000000000000000000000000000000000
```

**Comment by jack8086wang** [comment]
已经修改specs/dev.toml,   run dev chain始终会报下面的错误
2020-05-12 09:37:36.677 +08:00 main INFO sentry  sentry is disabled
2020-05-12 09:37:36.741 +08:00 main INFO main  Miner is disabled, edit ckb.toml to enable it
Run error: Error { kind: GenesisMismatch { expected: Byte32(0x823b2ff5785b12da8b1363cac9a5cbe566d8b715a4311441b119c39a0367488c), actual: Byte32(0x63547ecf6fc22d1325980c524b268b4a044d49cda3efbd584c0a8c8b9faaf9e1) }

stack backtrace:
   0: failure::backtrace::internal::InternalBacktrace::new
   1: <failure::backtrace::Backtrace as core::default::Default>::default
   2: ckb_chain_spec::error::<impl core::convert::From<ckb_chain_spec::error::SpecError> for ckb_error::Error>::from
   3: ckb_shared::shared::Shared::init
   4: ckb_shared::shared::SharedBuilder::build
   5: ckb_bin::subcommand::run::run
   6: ckb_bin::run_app
   7: ckb::main
   8: std::rt::lang_start::{{closure}}
   9: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
      std::panicking::try::do_call
             at src/libstd/panicking.rs:292
  10: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  11: std::panicking::try
             at src/libstd/panicking.rs:270
      std::panic::catch_unwind
             at src/libstd/panic.rs:394
      std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  12: main
  13: __libc_start_main
  14: _start

**Comment by doitian** [comment]
> 已经修改specs/dev.toml, run dev chain始终会报下面的错误
> 2020-05-12 09:37:36.677 +08:00 main INFO sentry sentry is disabled
> 2020-05-12 09:37:36.741 +08:00 main INFO main Miner is disabled, edit ckb.toml to enable it
> Run error: Error { kind: GenesisMismatch { expected: Byte32(0x823b2ff5785b12da8b1363cac9a5cbe566d8b715a4311441b119c39a0367488c), actual: Byte32(0x63547ecf6fc22d1325980c524b268b4a044d49cda3efbd584c0a8c8b9faaf9e1) }
> 
> stack backtrace:
> 0: failure::backtrace::internal::InternalBacktrace::new
> 1: <failure::backtrace::Backtrace as core::default::Default>::default
> 2: ckb_chain_spec::error::<impl core::convert::From<ckb_chain_spec::error::SpecError> for ckb_error::Error>::from
> 3: ckb_shared::shared::Shared::init
> 4: ckb_shared::shared::SharedBuilder::build
> 5: ckb_bin::subcommand::run::run
> 6: ckb_bin::run_app
> 7: ckb::main
> 8: std::rt::lang_start::{{closure}}
> 9: std::rt::lang_start_internal::{{closure}}
> at src/libstd/rt.rs:52
> std::panicking::try::do_call
> at src/libstd/panicking.rs:292
> 10: __rust_maybe_catch_panic
> at src/libpanic_unwind/lib.rs:78
> 11: std::panicking::try
> at src/libstd/panicking.rs:270
> std::panic::catch_unwind
> at src/libstd/panic.rs:394
> std::rt::lang_start_internal
> at src/libstd/rt.rs:51
> 12: main
> 13: __libc_start_main
> 14: _start

需要修改 ckb.toml 填入矿工地址

```
[block_assembler]
code_hash = "0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8"
args = "0x5406f5abe8e3...ba641f79556688c226d44631d"
hash_type = "type"
message = "0x"
```

args 改成 secp256k1 pubkey 的 blake2b hash

**Comment by jack8086wang** [comment]
修改ckb.toml和dev.toml后dev  ./ckb run还是一样的错误。
 [block_assembler]
 code_hash = "0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8"
 args = "0x321f602d0725f3103774cf877be83c91ed1027b0"
 hash_type = "type"
 message = "0x"
          
另外请帮忙在提交一下下面的rpc request看下校验结果是不是通过的，powhash<block target. 如果用这个提交校验还是能通过，我再找下问题可能的地方：
headerhash:              0x8ac19359f95118c2f08a5916645b728d565efaebfff4d3381f0e3da63fad307b
nonce:                       0xca7721790000000000001f9d689da
本地验算powhash:      34d3bbe9af07cd338f78a8c550bdf54a526ecf052a90188de6f0b4
blocktarget:                ab7a4700000000000000000000000000000000000000000000000000

rpc params =  ["0xc",{"header":{"version":"0x0","compact_target":"0x1cab7a47","timestamp":"0x1720734f4fb","number":"0x326d8","epoch":"0x2370059000117","transactions_root":"0x287e60627c9e285c6def13c16195ad52dce87c3e2c317cd47a1f6f57f7c95f06","proposals_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","uncles_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","parent_hash":"0xee1437bfb4718b5231d00e949bc6b34216a57d3c09cf526b1c9d01e7e4131bda","dao":"0x2e4798157a2f9c2fc7ed3f82f4b4230091c6b3afcfcc330000e0b536e04e0707","nonce":"0xca7721790000000000001f9d689da"},"uncles":[],"transactions":[{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x326d8"}],"outputs":[{"capacity":"0x5278f14e47","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x590000000c00000055000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000321f602d0725f3103774cf877be83c91ed1027b000000000"]}],"proposals":[]}]

submit_block rpc result: error= { code: -3, message: 'Header(Pow(InvalidNonce))' } , result =  undefined

**Comment by doitian** [comment]
```
pow hash: 0x745cbf48e94e6dc39744bab4f7ae27d2a6319cab0f81ab077da5fdcce8c0f273
nonce: 0xca7721790000000000001f9d689da
pow input: [
    0x74,
    0x5c,
    0xbf,
    0x48,
    0xe9,
    0x4e,
    0x6d,
    0xc3,
    0x97,
    0x44,
    0xba,
    0xb4,
    0xf7,
    0xae,
    0x27,
    0xd2,
    0xa6,
    0x31,
    0x9c,
    0xab,
    0xf,
    0x81,
    0xab,
    0x7,
    0x7d,
    0xa5,
    0xfd,
    0xcc,
    0xe8,
    0xc0,
    0xf2,
    0x73,
    0xda,
    0x89,
    0xd6,
    0xf9,
    0x1,
    0x0,
    0x0,
    0x0,
    0x0,
    0x0,
    0x90,
    0x17,
    0x72,
    0xa7,
    0xc,
    0x0,
]
pow output: [
    0xef,
    0x4a,
    0x7,
    0xe5,
    0x1c,
    0xb0,
    0x97,
    0x79,
    0xea,
    0x29,
    0x1f,
    0x97,
    0x60,
    0x96,
    0xe2,
    0x3,
    0xe,
    0xa5,
    0x7f,
    0x63,
    0xf4,
    0x91,
    0xad,
    0xcb,
    0xda,
    0xff,
    0xe2,
    0x2f,
    0xf1,
    0xab,
    0x65,
    0x8b,
]
pow output as big endian: 0xef4a07e51cb09779ea291f976096e2030ea57f63f491adcbdaffe22ff1ab658b
block_target: 0xab7a4700000000000000000000000000000000000000000000000000
```

还是错的，你有比对中间结果吗？看日志 pow hash 是错的

**Comment by doitian** [comment]
可以用 https://github.com/doitian/ckb-sdk-python 对比测试一下

    pip install -U ckb

There's also an example method to convert from block template to a block https://github.com/doitian/ckb-sdk-python/blob/master/ckb/block.py#L45

```
import ckb.header
import json

block = json.loads("""
{"header":{"version":"0x0","compact_target":"0x1cab7a47","timestamp":"0x1720734f4fb","number":"0x326d8","epoch":"0x2370059000117","transactions_root":"0x287e60627c9e285c6def13c16195ad52dce87c3e2c317cd47a1f6f57f7c95f06","proposals_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","uncles_hash":"0x0000000000000000000000000000000000000000000000000000000000000000","parent_hash":"0xee1437bfb4718b5231d00e949bc6b34216a57d3c09cf526b1c9d01e7e4131bda","dao":"0x2e4798157a2f9c2fc7ed3f82f4b4230091c6b3afcfcc330000e0b536e04e0707","nonce":"0xca7721790000000000001f9d689da"},"uncles":[],"transactions":[{"cell_deps":[],"header_deps":[],"inputs":[{"previous_output":{"index":"0xffffffff","tx_hash":"0x0000000000000000000000000000000000000000000000000000000000000000"},"since":"0x326d8"}],"outputs":[{"capacity":"0x5278f14e47","lock":{"args":"0xda648442dbb7347e467d1d09da13e5cd3a0ef0e1","code_hash":"0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8","hash_type":"type"},"type":null}],"outputs_data":["0x"],"version":"0x0","witnesses":["0x590000000c00000055000000490000001000000030000000310000009bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce80114000000321f602d0725f3103774cf877be83c91ed1027b000000000"]}],"proposals":[]}
""")
print(ckb.header.pow_hash(block['header']))
# => 0x745cbf48e94e6dc39744bab4f7ae27d2a6319cab0f81ab077da5fdcce8c0f273
```

**Comment by jack8086wang** [comment]
谢谢 doitian, 我之前的理解是错误的，之前eaglesonghash(headerHashWithNonce)< blocktarget,我就去做submit_block, 正确的方法应该是eaglesonghash(ckb.header.pow_hash(block['header'])+nonce)<blocktarget 再做submit

**Comment by jack8086wang** [comment]
dotian帮我确认一下我上述表述是不是正确的，现在我这样处理后测试矿机report上来的nonce值去校验全部>blocktarget..... (V0.31.1版本node的testnet)

**Comment by jack8086wang** [comment]
搞定了，再次感谢dotian
